src/util/notifications_util.js

```diff
@@ -0,0 +1,379 @@
+    }
+
+    destroy() {
+        this.connection.flush(10000);
+        this.connection.disconnect();
+    }
+}
+
+//replace properties starting with 'local_file'
+//with the content of the pointed file
+//(useful for loading tls certificates into http options object)
+function load_files(object) {
+    if (typeof object !== 'object') return;
+    for (const key in object) {
+        if (key.startsWith('local_file')) {
+            const new_key = key.substring('local_file_'.length);
+            const content = fs.readFileSync(object[key], 'utf-8');
+            object[new_key] = content;
+            delete object[key];
+        } else {
+            load_files(object[key]);
+        }
+    }
+    dbg.log2('load_files for obj =', object);
+}
+
+function parse_connect_file(connect_filepath) {
+    const connect = {};
+    const connect_strs = fs.readFileSync(connect_filepath, 'utf-8').split(os.EOL);
+    for (const connect_str of connect_strs) {
+        if (connect_str === '') continue;
+        const kv = connect_str.split('=');
+        //parse JSONs-
+        if (kv[0].endsWith('object')) {
+            kv[1] = JSON.parse(kv[1]);
+        }
+        connect[kv[0]] = kv[1];
+    }
+    //parse file contents (useful for tls cert files)
+    load_files(connect);
+    return connect;
+}
+
+
+function get_connection(connect) {
+    switch (connect.notification_protocol.toLowerCase()) {
+        case 'http':
+        case 'https':
+        {
+            return new HttpNotificator(connect);
+        }
+        case 'kafka': {
+            return new KafkaNotificator(connect);
+        }
+        default: {
+            dbg.error("Unknown notification protocol", connect.notification_protocol);
+            //nothing more to possibly do with this notification, don't append to failure log
+        }
+    }
+}
+
+
+async function test_notifications(bucket) {
+    for (const notif of bucket.notifications) {
+        const connect = parse_connect_file(notif.connect);
+        dbg.log1("testing notif", notif);
+        try {
+            const connection = get_connection(connect);
+            await connection.connect();
+            await connection.promise_notify({notif: "test notification"}, async err => err);
+            connection.destroy();
+        } catch (err) {
+            dbg.error("Connection failed for", connect);
+            return err;
+        }
+    }
+}
+
+//see https://docs.aws.amazon.com/AmazonS3/latest/userguide/notification-content-structure.html
+function compose_notification(req, res, bucket, notif_conf) {
+    let eTag = res.getHeader('ETag');
+    //eslint-disable-next-line
+    if (eTag && eTag.startsWith('\"') && eTag.endsWith('\"')) {
+        eTag = eTag.substring(2, eTag.length - 2);
+    }
+
+    const notif = {
+        eventVersion: '2.3',
+        eventSource: _get_system_name(req) + ':s3',
+        eventTime: new Date().toISOString,
+        eventName: req.s3event + ':' + (req.s3_event_op || req.method),
```

Comment:
I suggest instead of setting `req.s3event` in each op you can maintain an object to map `req.op_name` to the `s3event`. similar to `OP_NAME_TO_ACTION` in `s3_bucket_policy_utils.js`

Reply:
Ok.

This was done to s3event, the second element in s3 event notation.
For the third element, called method in my code, it was possible in some cases to use the OP_NAME_TO_ACTION object (eg post_object_uploadId), but some cases (eg DeleteMarkerCreated) have to be differentiated manually in code.

I've also changed terminology in the filtering code to (try to) improve readability.